File: code/java/DesignPatterns/src/main/java/iterator/PostIterator.java
package iterator;

import java.util.List;
import java.util.NoSuchElementException;

class PostIterator implements Iterator {
    private List<Post> posts;
    private int position;

    public PostIterator(List<Post> posts) {
        this.posts = posts;
        this.position = 0;
    }

    @Override
    public boolean hasNext() {
        return position < posts.size();
    }

    @Override
    public Post next() {
        if (!hasNext()) {
            throw new NoSuchElementException("No more posts available");
        }
        Post post = posts.get(position);
        position++;
        return post;
    }
}

File: code/csharp/DesignPatterns/iterator/PostIterator.cs
namespace Iterator;
using System;
using System.Collections.Generic;

class PostIterator : IIterator
{
    private List<Post> _posts;
    private int _position = 0;

    public PostIterator(List<Post> posts)
    {
        _posts = posts;
    }

    public bool HasNext()
    {
        return _position < _posts.Count;
    }

    public Post Next()
    {
        if (!HasNext())
        {
            throw new InvalidOperationException("No more posts available");
        }
        return _posts[_position++];
    }
}
